In [1]:
import numpy as np
import pandas as pd

In [2]:
health_df = pd.read_csv(r"D:\Imarticus_Data Science\Machine Learning\Linear Regression using OLS_Model\insurance.csv")
health_df.sample(5)

age     sex     bmi  children smoker     region      charges
563    50    male  44.770         1     no  southeast   9058.73030
1082   38    male  19.950         1     no  northwest   5855.90250
596    42  female  29.480         2     no  southeast   7640.30920
367    42  female  24.985         2     no  northwest   8017.06115
803    18  female  42.240         0    yes  southeast  38792.68560

In [3]:
health_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [4]:
health_df.describe(include='object')

sex smoker     region
count   1338   1338       1338
unique     2      2          4
top     male     no  southeast
freq     676   1064        364

In [5]:
health_df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

In [ ]:
health_df.skew()

Since the data has a normal distribustion , we use Feature scaling method to do Standardization

In [15]:
from sklearn import preprocessing, model_selection 
# from sklearn.preprocessing import StandardScaler

labelencoder = preprocessing.LabelEncoder()   
scaler = preprocessing.StandardScaler()  


In [16]:
health_df.isna().sum()        # Confirms , No NAN values 

age         0
sex         0
bmi         0
children    0
smoker      0
charges     0
dtype: int64

In [17]:
health_df.drop_duplicates(inplace=True)

C:\Users\Amulya Tandur\AppData\Local\Temp\ipykernel_31608\3792054197.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  health_df.drop_duplicates(inplace=True)


In [18]:
health_df = health_df[['age','sex','bmi','children','smoker','charges']]

### fit(), tranform() and fit_transform() -> vital tools that are extensively used in data preprocessing and model fitting.

In [19]:
# Apply label encoding to SEX and Smoker variables 
from sklearn.preprocessing import LabelEncoder

le_sex = LabelEncoder()
le_smoker = LabelEncoder()

# Encode safely with .loc[] & Fit-transform and assign with explicit dtype conversion
health_df.loc[:,'sex'] = le_sex.fit_transform(health_df['sex']).astype('int64')
health_df.loc[:,'smoker'] = le_smoker.fit_transform(health_df['smoker']).astype('int64')


In [20]:
health_df.head()

age sex     bmi  children smoker      charges
0   19   0  27.900         0      1  16884.92400
1   18   1  33.770         1      0   1725.55230
2   28   1  33.000         3      0   4449.46200
3   33   1  22.705         0      0  21984.47061
4   32   1  28.880         0      0   3866.85520

# Split X(independent variables) and Y(dependent variable)

In [21]:

X = health_df.drop(['charges'], axis =1)
y = health_df[['charges']]

In [22]:
X.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker'], dtype='object')

In [23]:
scaled_X =scaler.fit_transform(X)     

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(scaled_X, y,test_size =0.3, random_state =1)

In [25]:
print("The shape of X_train is:",X_train.shape)
print("The shape of X_test is:",X_test.shape)
print("The shape of y_train is:",y_train.shape)
print("The shape of y_test is:",y_test.shape)

The shape of X_train is: (935, 5)
The shape of X_test is: (402, 5)
The shape of y_train is: (935, 1)
The shape of y_test is: (402, 1)


In [26]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.formula.api as ols
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import SGDRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error,r2_score

import warnings
warnings.filterwarnings('ignore')

In [27]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [28]:
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)

Coefficients: [[3620.79229826 -263.89828818 2292.5531317   671.56343292 9440.44045578]]
Intercept: [13312.60990194]


In [29]:
print(model.score(X_train, y_train))

# This code in scikit-learn returns the coefficient of determination (R² score) by default for regression models.

0.7534801154526991


### An R² score of 0.7534 means that only 75.3% of the variance in the target variable (charges) using the features in X_train.

In [30]:
print(model.score(X_test, y_test))

0.7330441380622205


###  Step 4: Make Predictions

In [31]:
y_pred = model.predict(X_test)

### Step 5: Evaluate the Model

In [32]:
from statsmodels.tools.eval_measures import rmse

R2 =r2_score(y_test,y_pred)
print('R2 Score between y_test and predicted :', R2)

RMSE = rmse(y_test,y_pred)
print('RMSE between y_test and predicted :', RMSE)


R2 Score between y_test and predicted : 0.7330441380622205
RMSE between y_test and predicted : [6047.00743507]


### Steps to Deploy the Model with Streamlit 

In [33]:
import pickle

pickle.dump(model,open('insurance_model.pkl','wb'))

pickle.dump(scaler,open("scaler.pkl", "wb"))

In [1]:
!pip install streamlit

In [2]:
%%writefile app.py       

#Now, you will have a file called "app.py" saved in the working directory.

import numpy as np
import pandas as pd
import streamlit as st
import pickle 

model=pickle.load(open('insurance_model.pkl','rb'))
scaler = pickle.load(open("scaler.pkl", "rb"))


def preprocess_input(age,sex, bmi, children, smoker):
    data = pd.DataFrame({"age": [age],
                         "sex": [sex],
                         "bmi": [bmi],
                         "children":[children],
                         "smoker": [smoker] })
    return data



def predict_insurance_charges(data):
    data_scaled = scaler.transform(data)     
    prediction = model.predict(data_scaled)
    return prediction

# Main app function
def main():
    st.title('Insurance Charges Prediction App')
    st.write("Enter customer details to predict insurance charges")
    st.sidebar.title('User Input')
    
    age = st.sidebar.slider("Age",20,100, step =1,value =30)    
                                                                
    sex = st.sidebar.selectbox("Sex",[0,1], format_func =lambda x: "Female" if x==0 else "Male")
    bmi = st.sidebar.slider("BMI", 10.0, 40.0, step =0.1, value =20.0)
    children = st.sidebar.slider('Number of Children', 0,10, step =1, value =0)
    smoker = st.sidebar.selectbox("Smoker",[0,1], format_func =lambda x: "No" if x==0 else "Yes")

    if st.button("Predict Charges"):
        input_data = preprocess_input(age,sex, bmi, children, smoker)
        prediction = predict_insurance_charges(input_data)
        
        st.subheader('Estimated Insurance Charge :')
        st.success(f"${float(prediction[0]):,.2f}")        

if __name__ == "__main__":
    main()

Writing app.py


### Step 6: Run the app 